plugins/image-source/obs-slideshow.c

```diff
@@ -343,21 +425,161 @@ static void ss_update(void *data, obs_data_t *settings)
 			OBS_TRANSITION_SCALE_ASPECT);
 
 	if (ss->randomize && ss->files.num)
-		ss->cur_item = random_file(ss);
+		ss->cur_item = (int)random_file(ss);
 	if (new_tr)
 		obs_source_add_active_child(ss->source, new_tr);
 	if (ss->files.num)
-		obs_transition_start(ss->transition, OBS_TRANSITION_MODE_AUTO,
-				ss->tr_speed,
-				ss->files.array[ss->cur_item].source);
+		do_transition(ss, false);
 
 	obs_data_array_release(array);
 }
 
+static void next_slide(void *data)
+{
+	struct slideshow *ss = data;
+
+	if (++ss->cur_item >= (int)ss->files.num) {
+			ss->cur_item = 0;
+	}
+
+	do_transition(ss, false);
+}
+
+static void previous_slide(void *data)
+{
+	struct slideshow *ss = data;
+
+	if (--ss->cur_item <= 0) {
+		ss->cur_item = ss->files.num - 1;
+	}
+
+	do_transition(ss, false);
+}
+
+static void play_pause(void *data)
+{
+	struct slideshow *ss = data;
+
+	ss->paused = !ss->paused;
+	ss->manual = ss->paused;
+}
+
+static void restart(void *data)
+{
+	struct slideshow *ss = data;
+
+	ss->elapsed = 0.0f;
+	ss->cur_item = 0;
+
+	obs_transition_set(ss->transition,
+			ss->files.array[(size_t)ss->cur_item].source);
+
+	ss->stop = false;
+	ss->paused = false;
+}
+
+static void stop(void *data)
+{
+	struct slideshow *ss = data;
+
+	ss->elapsed = 0.0f;
+	ss->cur_item = 0;
+
+	do_transition(ss, true);
+	ss->stop = true;
+	ss->paused = false;
+}
+
+static void play_pause_hotkey(void *data, obs_hotkey_id id,
+		obs_hotkey_t *hotkey, bool pressed)
+{
+	UNUSED_PARAMETER(id);
+	UNUSED_PARAMETER(hotkey);
+
+	struct slideshow *ss = data;
+
+	if (pressed) {
+		play_pause(ss);
+	}
+}
+
+static void restart_hotkey(void *data, obs_hotkey_id id,
+		obs_hotkey_t *hotkey, bool pressed)
+{
+	UNUSED_PARAMETER(id);
+	UNUSED_PARAMETER(hotkey);
+
+	struct slideshow *ss = data;
+
+	if (pressed) {
+		restart(ss);
+	}
+}
+
+static void stop_hotkey(void *data, obs_hotkey_id id,
+		obs_hotkey_t *hotkey, bool pressed)
+{
+	UNUSED_PARAMETER(id);
+	UNUSED_PARAMETER(hotkey);
+
+	struct slideshow *ss = data;
+
+	if (pressed) {
+		stop(ss);
+	}
+}
+
+static void next_slide_hotkey(void *data, obs_hotkey_id id,
+		obs_hotkey_t *hotkey, bool pressed)
+{
+	UNUSED_PARAMETER(id);
+	UNUSED_PARAMETER(hotkey);
+
+	struct slideshow *ss = data;
+
+	if (!ss->manual)
+		return;
+
+	if (pressed) {
+		next_slide(ss);
+	}
+}
+
+static void previous_slide_hotkey(void *data, obs_hotkey_id id,
+		obs_hotkey_t *hotkey, bool pressed)
+{
+	UNUSED_PARAMETER(id);
+	UNUSED_PARAMETER(hotkey);
+
+	struct slideshow *ss = data;
+
+	if (!ss->manual)
+		return;
+
+	if (pressed) {
```

Comment:
Style:  This `if` statement does not need curly braces since its contents only use one line.